Data azurerm_servicebus_queue fails to be read with "cannot parse an empty string"

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->


### Terraform Version

1.2.5

### AzureRM Provider Version

3.15.0

### Affected Resource(s)/Data Source(s)

azurerm_servicebus_queue

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.15.0"
    }
  }
}

provider "azurerm" {
  features {}
}

data "azurerm_servicebus_queue" "status_messages" {
  name                = "<name of queue>"
  resource_group_name = "<resource group name>"   
  namespace_name      = "<namespace of service bus>"      
}
```


### Debug Output/Panic Output

```shell
data.azurerm_servicebus_queue.status_messages: Reading...
2022-07-25T09:31:59.231-0400 [ERROR] provider.terraform-provider-azurerm_v3.15.0_x5.exe: Response contains error diagnostic: tf_rpc=ReadDataSource @module=sdk.proto diagnostic_summary="parsing "": cannot parse an 
empty string" tf_provider_addr=provider diagnostic_detail= diagnostic_severity=ERROR tf_data_source_type=azurerm_servicebus_queue tf_proto_version=5.2 @caller=github.com/hashicorp/terraform-plugin-go@v0.10.0/tfprotov5/internal/diag/diagnostics.go:56 tf_req_id=3b44889c-d61c-00cd-5641-d946cd20c9b0 timestamp=2022-07-25T09:31:59.231-0400
2022-07-25T09:31:59.231-0400 [ERROR] vertex "data.azurerm_servicebus_queue.status_messages" error: parsing "": cannot parse an empty string
2022-07-25T09:31:59.233-0400 [ERROR] vertex "data.azurerm_servicebus_queue.status_messages" error: parsing "": cannot parse an empty string
2022-07-25T09:31:59.233-0400 [ERROR] vertex "data.azurerm_servicebus_queue.status_messages (expand)" error: parsing "": cannot parse an empty string
2022-07-25T09:31:59.233-0400 [INFO]  backend/local: plan operation completed
╷
│ Error: parsing "": cannot parse an empty string
│
│   with data.azurerm_servicebus_queue.status_messages,
│   on dependencies.tf line 2, in data "azurerm_servicebus_queue" "status_messages":
│    2: data "azurerm_servicebus_queue" "status_messages" {
│
╵
```


### Expected Behaviour

The existing resource is read correctly

### Actual Behaviour

The plan fails to complete with the message : "parsing "": cannot parse an empty string" 

### Steps to Reproduce

1. `terraform plan` on the snippet
2. Error will show up

Issue will show up whether the resource exits or not

### Important Factoids

_No response_

### References

_No response_

--- a/internal/services/postgres/migration/postgresql_server.go
+++ b/internal/services/postgres/migration/postgresql_server.go
@@ -1,0 +1,264 @@
+package migration
+
+import (
+	"context"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/postgresql/2017-12-01/servers"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/azure"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/postgres/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+)
+
+var _ pluginsdk.StateUpgrade = PostgresqlServerV0ToV1{}
+
+type PostgresqlServerV0ToV1 struct{}
+
+func (PostgresqlServerV0ToV1) Schema() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validate.ServerName,
+		},
+
+		"location": azure.SchemaLocation(),
+
+		"resource_group_name": azure.SchemaResourceGroupName(),
+
+		"sku_name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
+
+		"version": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringInSlice(servers.PossibleValuesForServerVersion(), false),
+		},
+
+		"administrator_login": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			Computed:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.All(validation.StringIsNotWhiteSpace, validate.AdminUsernames),
+		},
+
+		"administrator_login_password": {
+			Type:      pluginsdk.TypeString,
+			Optional:  true,
+			Sensitive: true,
+		},
+
+		"auto_grow_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+
+		"backup_retention_days": {
+			Type:         pluginsdk.TypeInt,
+			Optional:     true,
+			Computed:     true,
+			ValidateFunc: validation.IntBetween(7, 35),
+		},
+
+		"geo_redundant_backup_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			ForceNew: true,
+			Default:  false,
+		},
+
+		"create_mode": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			Default:      string(servers.CreateModeDefault),
+			ValidateFunc: validation.StringInSlice(servers.PossibleValuesForCreateMode(), false),
+		},
+
+		"creation_source_server_id": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ValidateFunc: servers.ValidateServerID,
+		},
+
+		"identity": commonschema.SystemAssignedIdentityOptional(),
+
+		"infrastructure_encryption_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			ForceNew: true,
+		},
+
+		"public_network_access_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+
+		"restore_point_in_time": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.IsRFC3339Time,
+		},
+
+		"storage_mb": {
+			Type:     pluginsdk.TypeInt,
+			Optional: true,
+			Computed: true,
+			ValidateFunc: validation.All(
+				validation.IntBetween(5120, 16777216),
+				validation.IntDivisibleBy(1024),
+			),
+		},
+
+		"ssl_minimal_tls_version_enforced": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			Default:      string(servers.MinimalTlsVersionEnumTLSOneTwo),
+			ValidateFunc: validation.StringInSlice(servers.PossibleValuesForMinimalTlsVersionEnum(), false),
+		},
+
+		"ssl_enforcement_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Required: true,
+		},
+
+		"threat_detection_policy": {
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"enabled": {
+						Type:     pluginsdk.TypeBool,
+						Optional: true,
+						AtLeastOneOf: []string{
+							"threat_detection_policy.0.enabled", "threat_detection_policy.0.disabled_alerts", "threat_detection_policy.0.email_account_admins",
+							"threat_detection_policy.0.email_addresses", "threat_detection_policy.0.retention_days", "threat_detection_policy.0.storage_account_access_key",
+							"threat_detection_policy.0.storage_endpoint",
+						},
+					},
+
+					"disabled_alerts": {
+						Type:     pluginsdk.TypeSet,
+						Optional: true,
+						Set:      pluginsdk.HashString,
+						Elem: &pluginsdk.Schema{
+							Type: pluginsdk.TypeString,
+							ValidateFunc: validation.StringInSlice([]string{
+								"Sql_Injection",
+								"Sql_Injection_Vulnerability",
+								"Access_Anomaly",
+								"Data_Exfiltration",
+								"Unsafe_Action",
+							}, false),
+						},
+						AtLeastOneOf: []string{
+							"threat_detection_policy.0.enabled", "threat_detection_policy.0.disabled_alerts", "threat_detection_policy.0.email_account_admins",
+							"threat_detection_policy.0.email_addresses", "threat_detection_policy.0.retention_days", "threat_detection_policy.0.storage_account_access_key",
+							"threat_detection_policy.0.storage_endpoint",
+						},
+					},
+
+					"email_account_admins": {
+						Type:     pluginsdk.TypeBool,
+						Optional: true,
+						AtLeastOneOf: []string{
+							"threat_detection_policy.0.enabled", "threat_detection_policy.0.disabled_alerts", "threat_detection_policy.0.email_account_admins",
+							"threat_detection_policy.0.email_addresses", "threat_detection_policy.0.retention_days", "threat_detection_policy.0.storage_account_access_key",
+							"threat_detection_policy.0.storage_endpoint",
+						},
+					},
+
+					"email_addresses": {
+						Type:     pluginsdk.TypeSet,
+						Optional: true,
+						Elem: &pluginsdk.Schema{
+							Type: pluginsdk.TypeString,
+							// todo email validation in code
+						},
+						Set: pluginsdk.HashString,
+						AtLeastOneOf: []string{
+							"threat_detection_policy.0.enabled", "threat_detection_policy.0.disabled_alerts", "threat_detection_policy.0.email_account_admins",
+							"threat_detection_policy.0.email_addresses", "threat_detection_policy.0.retention_days", "threat_detection_policy.0.storage_account_access_key",
+							"threat_detection_policy.0.storage_endpoint",
+						},
+					},
+
+					"retention_days": {
+						Type:         pluginsdk.TypeInt,
+						Optional:     true,
+						ValidateFunc: validation.IntAtLeast(0),
+						AtLeastOneOf: []string{
+							"threat_detection_policy.0.enabled", "threat_detection_policy.0.disabled_alerts", "threat_detection_policy.0.email_account_admins",
+							"threat_detection_policy.0.email_addresses", "threat_detection_policy.0.retention_days", "threat_detection_policy.0.storage_account_access_key",
+							"threat_detection_policy.0.storage_endpoint",
+						},
+					},
+
+					"storage_account_access_key": {
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
+						Sensitive:    true,
+						ValidateFunc: validation.StringIsNotEmpty,
+						AtLeastOneOf: []string{
+							"threat_detection_policy.0.enabled", "threat_detection_policy.0.disabled_alerts", "threat_detection_policy.0.email_account_admins",
+							"threat_detection_policy.0.email_addresses", "threat_detection_policy.0.retention_days", "threat_detection_policy.0.storage_account_access_key",
+							"threat_detection_policy.0.storage_endpoint",
+						},
+					},
+
+					"storage_endpoint": {
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
+						ValidateFunc: validation.StringIsNotEmpty,
+						AtLeastOneOf: []string{
+							"threat_detection_policy.0.enabled", "threat_detection_policy.0.disabled_alerts", "threat_detection_policy.0.email_account_admins",
+							"threat_detection_policy.0.email_addresses", "threat_detection_policy.0.retention_days", "threat_detection_policy.0.storage_account_access_key",
+							"threat_detection_policy.0.storage_endpoint",
+						},
+					},
+				},
+			},
+		},
+
+		"fqdn": {
+			Type:     pluginsdk.TypeString,
+			Computed: true,
+		},
+
+		"tags": commonschema.Tags(),
+	}
+}
+
+func (PostgresqlServerV0ToV1) UpgradeFunc() pluginsdk.StateUpgraderFunc {
+	return func(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+		// old
+		//  /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBForPostgreSQL/servers/{serverName}
+		// new:
+		//  /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforPostgreSQL/servers/{serverName}
+		// summary:
+		// Check for `For` and swap to `for`
+		oldId := rawState["id"].(string)
+		if strings.Contains(oldId, "Microsoft.DBForPostgreSQL") {
+			modifiedId := strings.ReplaceAll(oldId, "Microsoft.DBForPostgreSQL", "Microsoft.DBforPostgreSQL")
+
+			newId, err := servers.ParseServerID(modifiedId)
+			if err != nil {
+				return rawState, err
+			}
+			log.Printf("[DEBUG] Updating ID from %q to %q", oldId, newId)
+			rawState["id"] = newId.ID()
+		}
+
+		return rawState, nil
+	}
+}

--- a/internal/services/postgres/postgresql_server_resource.go
+++ b/internal/services/postgres/postgresql_server_resource.go
@@ -8,6 +8,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/postgres/migration"
+
 	"github.com/hashicorp/go-azure-helpers/lang/response"
 	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
 	"github.com/hashicorp/go-azure-helpers/resourcemanager/identity"
@@ -101,6 +103,11 @@ func resourcePostgreSQLServer() *pluginsdk.Resource {
 			Update: pluginsdk.DefaultTimeout(60 * time.Minute),
 			Delete: pluginsdk.DefaultTimeout(60 * time.Minute),
 		},
+
+		SchemaVersion: 1,
+		StateUpgraders: pluginsdk.StateUpgrades(map[int]pluginsdk.StateUpgrade{
+			0: migration.PostgresqlServerV0ToV1{},
+		}),
 
 		Schema: map[string]*pluginsdk.Schema{
 			"name": {

--- a/internal/services/servicebus/servicebus_queue_data_source.go
+++ b/internal/services/servicebus/servicebus_queue_data_source.go
@@ -37,18 +37,22 @@ func dataSourceServiceBusQueue() *pluginsdk.Resource {
 				AtLeastOneOf: []string{"namespace_id", "resource_group_name", "namespace_name"},
 			},
 
+			// TODO Remove in 4.0
 			"namespace_name": {
 				Type:         pluginsdk.TypeString,
 				Optional:     true,
 				ValidateFunc: validate.NamespaceName,
 				AtLeastOneOf: []string{"namespace_id", "resource_group_name", "namespace_name"},
+				Deprecated:   "`namespace_name` will be removed in favour of the property `namespace_id` in version 4.0 of the AzureRM Provider.",
 			},
 
+			// TODO Remove in 4.0
 			"resource_group_name": {
 				Type:         pluginsdk.TypeString,
 				Optional:     true,
 				ValidateFunc: resourcegroups.ValidateName,
 				AtLeastOneOf: []string{"namespace_id", "resource_group_name", "namespace_name"},
+				Deprecated:   "`resource_group_name` will be removed in favour of the property `namespace_id` in version 4.0 of the AzureRM Provider.",
 			},
 
 			"auto_delete_on_idle": {
@@ -135,14 +139,24 @@ func dataSourceServiceBusQueue() *pluginsdk.Resource {
 func dataSourceServiceBusQueueRead(d *pluginsdk.ResourceData, meta interface{}) error {
 	client := meta.(*clients.Client).ServiceBus.QueuesClient
 	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
 	defer cancel()
 
-	namespaceId, err := namespaces.ParseNamespaceID(d.Get("namespace_id").(string))
-	if err != nil {
-		return err
+	var resourceGroup string
+	var namespaceName string
+	if v, ok := d.Get("namespace_id").(string); ok && v != "" {
+		namespaceId, err := namespaces.ParseNamespaceID(d.Get("namespace_id").(string))
+		if err != nil {
+			return err
+		}
+		resourceGroup = namespaceId.ResourceGroupName
+		namespaceName = namespaceId.NamespaceName
+	} else {
+		resourceGroup = d.Get("resource_group_name").(string)
+		namespaceName = d.Get("namespace_name").(string)
 	}
 
-	id := queues.NewQueueID(namespaceId.SubscriptionId, namespaceId.ResourceGroupName, namespaceId.NamespaceName, d.Get("name").(string))
+	id := queues.NewQueueID(subscriptionId, resourceGroup, namespaceName, d.Get("name").(string))
 
 	resp, err := client.Get(ctx, id)
 	if err != nil {
@@ -178,7 +192,8 @@ func dataSourceServiceBusQueueRead(d *pluginsdk.ResourceData, meta interface{}) 
 				// then the max size returned by the API will be 16 times greater than the value set.
 				if *props.EnablePartitioning {
 					namespacesClient := meta.(*clients.Client).ServiceBus.NamespacesClient
-					namespace, err := namespacesClient.Get(ctx, *namespaceId)
+					namespaceId := namespaces.NewNamespaceID(subscriptionId, resourceGroup, namespaceName)
+					namespace, err := namespacesClient.Get(ctx, namespaceId)
 					if err != nil {
 						return err
 					}

--- a/internal/services/servicebus/servicebus_queue_data_source_test.go
+++ b/internal/services/servicebus/servicebus_queue_data_source_test.go
@@ -37,6 +37,33 @@ func TestAccDataSourceServiceBusQueue_basic(t *testing.T) {
 	})
 }
 
+func TestAccDataSourceServiceBusQueue_withNamespaceName(t *testing.T) {
+	data := acceptance.BuildTestData(t, "data.azurerm_servicebus_queue", "test")
+	r := ServiceBusQueueDataSource{}
+
+	data.DataSourceTest(t, []acceptance.TestStep{
+		{
+			Config: r.withNamespaceName(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).Key("id").Exists(),
+				check.That(data.ResourceName).Key("auto_delete_on_idle").Exists(),
+				check.That(data.ResourceName).Key("dead_lettering_on_message_expiration").Exists(),
+				check.That(data.ResourceName).Key("default_message_ttl").Exists(),
+				check.That(data.ResourceName).Key("duplicate_detection_history_time_window").Exists(),
+				check.That(data.ResourceName).Key("enable_batched_operations").Exists(),
+				check.That(data.ResourceName).Key("enable_express").Exists(),
+				check.That(data.ResourceName).Key("enable_partitioning").Exists(),
+				check.That(data.ResourceName).Key("lock_duration").Exists(),
+				check.That(data.ResourceName).Key("max_delivery_count").Exists(),
+				check.That(data.ResourceName).Key("max_size_in_megabytes").Exists(),
+				check.That(data.ResourceName).Key("requires_duplicate_detection").Exists(),
+				check.That(data.ResourceName).Key("requires_session").Exists(),
+				check.That(data.ResourceName).Key("status").Exists(),
+			),
+		},
+	})
+}
+
 func (ServiceBusQueueDataSource) basic(data acceptance.TestData) string {
 	return fmt.Sprintf(`
 %s
@@ -44,6 +71,18 @@ func (ServiceBusQueueDataSource) basic(data acceptance.TestData) string {
 data "azurerm_servicebus_queue" "test" {
   name         = azurerm_servicebus_queue.test.name
   namespace_id = azurerm_servicebus_namespace.test.id
+}
+`, ServiceBusQueueResource{}.basic(data))
+}
+
+func (ServiceBusQueueDataSource) withNamespaceName(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_servicebus_queue" "test" {
+  name                = azurerm_servicebus_queue.test.name
+  namespace_name      = azurerm_servicebus_namespace.test.name
+  resource_group_name = azurerm_resource_group.test.name
 }
 `, ServiceBusQueueResource{}.basic(data))
 }

